Source/WebCore/editing/MarkupAccumulator.cpp

```diff
@@ -217,18 +212,39 @@ void MarkupAccumulator::serializeNodesWithNamespaces(Node& targetNode, Serialize
         namespaceHash.set(XMLNames::xmlNamespaceURI->impl(), xmlAtom().impl());
     }
 
-    if (root == SerializedNodes::SubtreeIncludingNode)
-        startAppendingNode(targetNode, &namespaceHash);
-
-    if (targetNode.document().isHTMLDocument() && elementCannotHaveEndTag(targetNode))
-        return;
-
-    Node* current = targetNode.hasTagName(templateTag) ? downcast<HTMLTemplateElement>(targetNode).content().firstChild() : targetNode.firstChild();
-    for ( ; current; current = current->nextSibling())
-        serializeNodesWithNamespaces(*current, SerializedNodes::SubtreeIncludingNode, &namespaceHash, tagNamesToSkip);
+    const Node* current = &targetNode;
+    do {
+        bool shouldTraverseChildren = true;
+        if (tagNamesToSkip && is<Element>(current)) {
+            for (auto& name : *tagNamesToSkip) {
+                if (downcast<Element>(current)->hasTagName(name))
+                    shouldTraverseChildren = false;
+            }
+        }
```

Comment:
I would move this computation after the shouldAppendNode computation, so it's closer to its use.